src/dali/abstract_pair_converter_plugin.py

```diff
@@ -88,11 +109,61 @@ def get_or_set_vertex(self, name: str) -> Vertex[ValueType]:
         self.add_vertex(new_vertex)
         return new_vertex
 
-    def add_neighbor(self, vertex_name: str, neighbor_name: str, weight: float = 0.0) -> None:
+    def is_optimized(self, asset: str) -> bool:
+        return asset in self.__optimized_assets
+
+    @property
+    def optimized_assets(self) -> Set[str]:
+        return self.__optimized_assets
+
+    # Optimization contains a dict with a key of the optimized asset and a value of a dict with the optimized weights for each neighbor
```

Comment:
Can optimizations be updated? Can you add a description of how that works?

Reply:
Optimizations are only needed for the cloning process. They are basically immutable, but at line 686 of `ccxt.py` the volumes are changed to weights.

We talked earlier about keeping things immutable and I attempted that here. I guess to make it more immutable, the function to add an optimized neighbor `add_neighbor()` needs to be made private.

Reply:
Immutability is fine but I was thinking of the case in which an optimization might become outdated (e.g. volume data changes, a market disappears and a new route is needed, etc.). Is there a way to recompute optimizations is such a case (this may require to clone a new graph, but that's fine). In other words: how final is it when an asset appears in __optimized_assets?

Reply:
I think there is a misunderstanding about how the optimization process works. Currently, the CCXT plugin uses the manifest to walk through week by week the volume of the exchanges and weighing the paths based on volume. These weeks are then joined together if they have the same ranking.

For example, we optimize for BTC -> USD on an exchange that doesn't have USD, so we must route through a stable coin:

week 1
BTC -> USDT volume is 100
BTC -> USDC volume is 10
week 2
BTC -> USDT volume is 100
BTC -> USDC volume is 20
week 3
BTC -> USDT volume is 10
BTC -> USDC volume is 200

When a transaction comes up requesting the price of BTC, CCXT will begin to optimize the graph snapshots for that exchange. It will make a snapshot of the graph for week 1 and 2 and place it in the AVLTree with a timestamp for week 1. There is no snapshot for week 2 since it would be identical. A new snapshot __is__ made for week 3 and placed in the AVLTree at week 3 timestamp.

Once these snapshots are made, currently, they are not changed or edited. They are used for the entire pricing process. Since the past can't change, there is no need to reoptimize a graph snapshot once it is made. Does that make sense?

Reply:
Yeah, I've been holding off until I can make sure it works, haha. I **think** this will give us better pricing, once I pull it off. When everything gets ironed out, I'll try to write a coherent comment or add it to the docs.